src/mesh/region_set.cc

```diff
@@ -0,0 +1,291 @@
+		.close();
+}
+
+
+
+const int RegionSetFromId::registrar =
+		Input::register_class< RegionSetFromId, const Input::Record &, Mesh * >("From_Id") +
+		RegionSetFromId::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetFromLabel
+ */
+
+RegionSetFromLabel::RegionSetFromLabel(const Input::Record &rec, Mesh *mesh)
+{
+	string region_name;
+	string mesh_label = rec.val<string>("mesh_label");
+	if ( !rec.opt_val<string>("name", region_name) ) {
+		region_name = mesh_label;
+	}
+
+	Region reg = mesh->region_db().find_label(mesh_label);
+	if (reg == Region()) {
+		xprintf(Warn, "Unknown region in mesh with label '%s'\n", mesh_label.c_str());
+	} else {
+		unsigned int region_id = reg.id();
+		this->add_region(mesh, region_id, region_name);
+	}
+}
+
+
+
+const IT::Record & RegionSetFromLabel::get_region_input_type()
+{
+    return IT::Record("From_Label", "Region declared by mesh_label and name.")
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(),
+				"Label (name) of the region. Has to be unique in one mesh.\n")
+		.declare_key("mesh_label", IT::String(), IT::Default::obligatory(),
+				"The mesh_label is e.g. physical volume name in GMSH format.")
+		.close();
+}
+
+
+
+const int RegionSetFromLabel::registrar =
+		Input::register_class< RegionSetFromLabel, const Input::Record &, Mesh * >("From_Label") +
+		RegionSetFromLabel::get_region_input_type().size();
+
+
+
+/*******************************************************************
+ * implementation of RegionSetFromElements
+ */
+
+RegionSetFromElements::RegionSetFromElements(const Input::Record &rec, Mesh *mesh)
+{
+	unsigned int region_id;
+	string region_label = rec.val<string>("name");
+
+	Input::Iterator<unsigned int> it = rec.find<unsigned int>("id");
+	if (it) {
+		region_id = (*it);
+	} else {
+		Region reg = mesh->region_db().find_label(region_label);
+		if ( reg.is_valid() ) region_id = reg.id();
+		else THROW( ExcNonexistingLabel() << EI_Region_Label(region_label) );
+	}
+
+	this->add_region(mesh, region_id, region_label);
+
+    Input::Array element_list;
+	if (rec.opt_val("element_list", element_list) ) {
+		std::vector<unsigned int> element_ids;
+		for (Input::Iterator<unsigned int> it_element = element_list.begin<unsigned int>();
+				it_element != element_list.end();
+		        ++it_element) {
+			element_ids.push_back( *it_element );
+
+		}
+		mesh->modify_element_ids(region_id, element_ids);
+	}
+}
+
+
+
+const IT::Record & RegionSetFromElements::get_region_input_type()
+{
+    return IT::Record("From_Elements", "Region declared by name and enum of elements.")
+        .derive_from(RegionSetBase::get_input_type())
+		.declare_key("name", IT::String(), IT::Default::obligatory(),
+				"Label (name) of the region. Has to be unique in one mesh.\n")
+		.declare_key("id", IT::Integer(0), IT::Default::optional(),
```

Comment:
Doc: "If not set the ID is set automatically to be unique. ID is used in output mesh, output field "region_id" and some error messages."